framework/aster-frame/src/arch/x86/mod.rs

```diff
@@ -99,3 +99,59 @@ fn enable_common_cpu_features() {
         });
     }
 }
+
+/// Copies `count * size_of::<T>()` bytes from `src` to `dst`.
+/// The source and destination may overlap.
+///
+/// If the source and destination will never overlap, `copy_nonoverlapping` can be used instead.
+///
+/// # Safety
+///
+/// The safety requirements of this function are consistent with `core::ptr::copy`.
+#[inline]
+pub unsafe fn copy<T>(src: *const T, dst: *mut T, count: usize) {
+    if src == dst || count == 0 {
+        return;
+    }
+
+    let is_overlapping = {
+        let src_end = src.add(count);
+        let dst_end = dst.add(count);
+        src < dst_end && dst < src_end as *mut T
+    };
+
+    if !is_overlapping {
+        copy_nonoverlapping(src, dst, count);
+    } else {
+        core::ptr::copy(src, dst, count);
+    }
+}
+
+/// Copies `count * size_of::<T>()` bytes from `src` to `dst`.
+/// The source and destination must not overlap.
+///
+/// For regions of memory which might overlap, use `copy` instead.
+///
+/// # Safety
+///
+/// The safety requirements of this function are consistent with `core::ptr::copy_nonoverlapping`.
+#[inline]
+pub unsafe fn copy_nonoverlapping<T>(src: *const T, dst: *mut T, count: usize) {
+    if count == 0 {
+        return;
+    }
+
+    let bytes_count = count * core::mem::size_of::<T>();
+    core::arch::asm!(
+        "push rcx",
+        "push rsi",
+        "push rdi",
+        "rep movsb",
```

Comment:
The `movsq` instruction is supposed to be more efficient than `movsb` if `count % 8 == 0`. Should we prefer `movsq` over `movsb`?

Reply:
```rust
 // In most cases, `movsq` is faster than `movsb`
    // because it transfers larger chunks of data in a single operation.
    if bytes_count % 8 == 0 && src.is_aligned_to(8) && dst.is_aligned_to(8) {
        core::arch::asm!(
            "rep movsq",
            in("rcx") bytes_count / 8,
            in("rsi") src,
            in("rdi") dst,
            lateout("rcx") _,
            lateout("rsi") _,
            lateout("rdi") _
        );
    } 
```